pkg/build/ignore/docker/dockign_test.go

```diff
@@ -0,0 +1,112 @@
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/openshift/source-to-image/pkg/api"
+	"os"
+)
+
+func newIgnorer() *DockIgn {
+	i := &DockIgn{
+	}
+	return i
+}
+
+func baseTest(t *testing.T, patterns []string, filesToCheck []string) {
+	cleanerr := os.RemoveAll("/tmp/upload")
+	if cleanerr != nil {
+		t.Errorf("problem cleaning up %v \n", cleanerr)
+	}
+	
+	c := &api.Config{WorkingDir: "/tmp"}
+	// create temp dir for .s2iignore that matches where dockin.go looks
+	dpath := filepath.Join(c.WorkingDir,"upload")
+	derr := os.Mkdir(dpath,0777)	
+	if derr != nil {
+		t.Errorf("Problem creating test dir %s with  %v \n", dpath, derr)
+	}
+	dpath = filepath.Join(dpath,"src")
+	derr = os.Mkdir(dpath,0777)
+	if derr != nil {
+		t.Errorf("Problem creating test dir %s with  %v \n", dpath, derr)
+	}
+	c.WorkingSourceDir = dpath
+	t.Logf("working source dir %s \n", dpath)
+
+	// create s2iignore file
+	ipath := filepath.Join(dpath, api.DefaultIgnoreFile)
+	ifile, ierr := os.Create(ipath)
+	defer ifile.Close()
+	if ierr != nil {
+		t.Errorf("Problem creating .s2iignore at %s with  %v \n",ipath, ierr)
+	}
+	// write files to remove into s2ignore, but save ! exclusions
+	filesToKeep := make(map[string]string)
+	for _,pattern := range patterns {
+		t.Logf("storing pattern %s \n",pattern)
+		_, serr := ifile.WriteString(pattern)
+		
+		if serr != nil {
+			t.Errorf("Problem setting .s2iignore %v \n", serr)
+		}
+		if strings.HasPrefix(pattern, "!") {
+			pattern = strings.Replace(pattern, "!","",1)
+			t.Logf("Should keep file %s \n", pattern)
+			filesToKeep[pattern] = pattern
+		}
+	}
+
+	// create file to delete
+        for _, fileToCreate := range filesToCheck {
+		fbpath := filepath.Join(dpath, fileToCreate)
+		fbfile, fberr := os.Create(fbpath)
+		defer fbfile.Close()
+		if fberr != nil {
+			t.Errorf("Problem opening test file %v \n", fberr)
+		}
+        }
+	
+	ignorer := newIgnorer()
+	
+	ignorer.Ignore(c)
+
+	for _,fileToCheck := range filesToCheck {
+		fbpath := filepath.Join(dpath, fileToCheck)
+		t.Logf("Evaluating delete cand %s \n", fbpath)
+		_, ok := filesToKeep[filepath.Base(fileToCheck)]
+		if ok {
+			t.Logf("skipping file to keep %s \n ", fileToCheck)
+			continue
+		}
+		ofile, oerr := os.Open(fbpath)
+		defer ofile.Close()
+		if oerr == nil {
+			t.Errorf("File was opened successfully when it should not have been %s \n", fbpath)
+		} else {
+			if !os.IsNotExist(oerr) {
+				t.Errorf("File %s could not be opened for an  error other than not existing,  %v \n", fbpath, oerr)
+			}
+		}
+	}
+
+	cleanerr = os.RemoveAll("/tmp/upload")
+	if cleanerr != nil {
+		t.Errorf("problem cleaning up %v \n", cleanerr)
+	}
+}
+
+func TestSingleIgnore(t *testing.T) {
+	baseTest(t, []string{"foo.bar\n"}, []string{"foo.bar"})
+}
+
+func TestWildcardIgnore(t *testing.T) {
+	baseTest(t, []string{"foo.*\n"}, []string{"foo.a", "foo.b"})
+}
+
+func TestExclusion(t *testing.T) {
+	baseTest(t, []string{"foo.*\n", "!foo.a"}, []string{"foo.a","foo.b"})
+}
```

Comment:
how about a test for files in subdirectories?
tests to ensure files not mentioned (either as include or excludes) at all are left around?

tests for ordering of !  (if file is mentioned before ! it gets retained, if file is mentioned after !, it is deleted, right?)

! with wildcard?

this is where it might be easer to create a couple test directory structures and just run your code against them, rather than building it programmatically.